File: src/main/scala/lox/Scanner.scala
package lox

import scala.collection.mutable.ListBuffer

object Scanner {
  private[lox] val Keywords: Map[String, TokenType] = Map(
    "and" -> TokenType.And,
    "class" -> TokenType.Class,
    "else" -> TokenType.Else,
    "false" -> TokenType.False,
    "for" -> TokenType.For,
    "fun" -> TokenType.Fun,
    "if" -> TokenType.If,
    "nil" -> TokenType.Nil,
    "or" -> TokenType.Or,
    "print" -> TokenType.Print,
    "return" -> TokenType.Return,
    "super" -> TokenType.Super,
    "this" -> TokenType.This,
    "true" -> TokenType.True,
    "var" -> TokenType.Var,
    "while" -> TokenType.While,
  )
}

final class Scanner(private val source: String) {
  final private val tokens = new ListBuffer[Token]
  private var start = 0
  private var current = 0
  private var line = 1

  private def reachedEnd(): Boolean = current >= source.length

  def scanTokens(): List[Token] = {
    while (!reachedEnd()) {
      // We are at the beginning of the next lexeme.
      start = current
      scanToken()
    }
    tokens += Token(TokenType.Eof, "", null, line)
    tokens.result()
  }

  private def scanToken(): Unit =
    advance() match {
      case '(' => addToken(TokenType.LeftParen)
      case ')' => addToken(TokenType.RightParen)
      case '{' => addToken(TokenType.LeftBrace)
      case '}' => addToken(TokenType.RightBrace)
      case ',' => addToken(TokenType.Comma)
      case '.' => addToken(TokenType.Dot)
      case '-' => addToken(TokenType.Minus)
      case '+' => addToken(TokenType.Plus)
      case ';' => addToken(TokenType.Semicolon)
      case '*' => addToken(TokenType.Star)
      case '!' if followedBy('=') => addToken(TokenType.BangEqual)
      case '!' => addToken(TokenType.Bang)
      case '=' if followedBy('=') => addToken(TokenType.EqualEqual)
      case '=' => addToken(TokenType.Equal)
      case '>' if followedBy('=') => addToken(TokenType.GreaterEqual)
      case '>' => addToken(TokenType.Greater)
      case '<' if followedBy('=') => addToken(TokenType.LessEqual)
      case '<' => addToken(TokenType.Less)
      case '/' if followedBy('/') => skipUntil('\n')
      case '/' => addToken(TokenType.Slash)
      case ' ' | '\r' | '\t' => // ignore whitespace
      case '\n' => line += 1
      case '"' => string()
      case n if isDigit(n) => number()
      case a if isAlpha(a) => identifier()
      case _ => Main.error(line, "Unexpected character.")
    }

  private def identifier(): Unit = {
    while (isAlphaNumeric(peek())) advance()
    val text = source.substring(start, current)
    addToken(Scanner.Keywords.getOrElse(text, TokenType.Identifier))
  }

  private def string(): Unit = {
    while ((peek() != '"') && !reachedEnd()) {
      if (peek() == '\n') line += 1
      advance()
    }
    if (reachedEnd()) {
      Main.error(line, "Unterminated string.")
      return
    }
    advance() // The closing quote
    val value = source.substring(start + 1, current - 1) // Trim the quotes
    addToken(TokenType.String, value)
  }

  private def number(): Unit = {
    while (isDigit(peek())) advance()
    // Look for a fractional part.
    if ((peek() == '.') && isDigit(peekNext())) {
      advance() // Consume the "."
      while (isDigit(peek())) advance()
    }
    addToken(TokenType.Number, source.substring(start, current).toDouble)
  }

  private def followedBy(expected: Char): Boolean = {
    if (reachedEnd()) return false
    if (source.charAt(current) != expected) return false
    current += 1
    true
  }

  private def skipUntil(char: Char): Unit =
    while (peek() != char && !reachedEnd()) advance()

  private def peek(): Char = {
    if (reachedEnd()) return '\u0000'
    source.charAt(current)
  }

  private def peekNext(): Char = {
    if (current + 1 >= source.length) return '\u0000'
    source.charAt(current + 1)
  }

  private def isAlpha(c: Char) =
    (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || c == '_'

  private def isAlphaNumeric(c: Char) = isAlpha(c) || isDigit(c)

  private def isDigit(c: Char) = c >= '0' && c <= '9'

  private def advance() = {
    val next = source.charAt(current)
    current += 1
    next
  }

  private def addToken(tokenType: TokenType): Unit =
    addToken(tokenType, null)

  private def addToken(tokenType: TokenType, literal: Any): Unit = {
    val text = source.substring(start, current)
    tokens += Token(tokenType, text, literal, line)
  }

}

File: lox/scanner.go
package lox

import (
	"bufio"
	"errors"
	"fmt"
	"io"
	"strconv"
	"unicode"
	"unicode/utf8"
)

type Scanner struct {
	reader  *bufio.Reader
	chars   []rune
	current int
	line    int
}

type syntaxError struct {
	line    int
	message string
}

func (e syntaxError) Error() string {
	return fmt.Sprintf("syntax error on line %d: %s", e.line, e.message)
}

func NewScanner(reader *bufio.Reader) *Scanner {
	return &Scanner{reader: reader, line: 1}
}

func (s *Scanner) NextToken() (Token, error) {
next:
	s.chars = s.chars[s.current:]
	s.current = 0
	r := s.advance()
	switch {
	case r == utf8.RuneError:
		return s.mkToken(EOF), nil
	case r == '(':
		return s.mkToken(LEFT_PAREN), nil
	case r == ')':
		return s.mkToken(RIGHT_PAREN), nil
	case r == '{':
		return s.mkToken(LEFT_BRACE), nil
	case r == '}':
		return s.mkToken(RIGHT_BRACE), nil
	case r == ',':
		return s.mkToken(COMMA), nil
	case r == '.':
		return s.mkToken(DOT), nil
	case r == '-':
		return s.mkToken(MINUS), nil
	case r == '+':
		return s.mkToken(PLUS), nil
	case r == ';':
		return s.mkToken(SEMICOLON), nil
	case r == '*':
		return s.mkToken(STAR), nil
	case r == '!':
		if s.match('=') {
			return s.mkToken(BANG_EQUAL), nil
		} else {
			return s.mkToken(BANG), nil
		}
	case r == '=':
		if s.match('=') {
			return s.mkToken(EQUAL_EQUAL), nil
		} else {
			return s.mkToken(EQUAL), nil
		}
	case r == '<':
		if s.match('=') {
			return s.mkToken(LESS_EQUAL), nil
		} else {
			return s.mkToken(LESS), nil
		}
	case r == '>':
		if s.match('=') {
			return s.mkToken(GREATER_EQUAL), nil
		} else {
			return s.mkToken(GREATER), nil
		}
	case r == '/':
		if s.match('/') {
			s.skipUntil(func(r rune) bool { return r == '\n' })
			goto next
		} else {
			return s.mkToken(SLASH), nil
		}
	case unicode.IsSpace(r):
		if r == '\n' {
			s.line += 1
		}
		goto next
	case r == '"':
		return s.str(), nil
	case unicode.IsDigit(r):
		return s.num()
	case unicode.IsLetter(r):
		return s.id(), nil
	default:
		return s.mkToken(ERROR), &syntaxError{s.line, "Unexpected character."}
	}
}

func (s *Scanner) id() Token {
	s.skipUntil(func(r rune) bool { return !unicode.IsDigit(r) && !unicode.IsLetter(r) })
	t, ok := keywords[string(s.chars[:s.current])]
	if !ok {
		t = IDENTIFIER
	}
	return s.mkToken(t)
}

func (s *Scanner) num() (Token, error) {
	s.skipUntil(func(r rune) bool { return !unicode.IsDigit(r) })
	// Look for a fractional part.
	if unicode.IsDigit(s.readRuneAhead(1)) && s.readRune() == '.' {
		s.current += 1 // skip the dot
		s.skipUntil(func(r rune) bool { return !unicode.IsDigit(r) })
	}
	if x, err := strconv.ParseFloat(string(s.chars[:s.current]), 64); err != nil {
		return s.mkToken(ERROR), &syntaxError{s.line, "Invalid number."}
	} else {
		return s.mkLiteral(NUMBER, x), nil
	}
}

func (s *Scanner) str() Token {
	s.current += 1 // skip the opening quote
	s.skipUntil(func(r rune) bool { return r == '"' })
	value := string(s.chars[1:s.current])
	s.current += 1 // skip the closing quote
	return s.mkLiteral(STRING, value)
}

func (s *Scanner) match(expected rune) bool {
	r := s.readRune()
	if r != expected {
		return false
	}
	s.current += 1
	return true
}

func (s *Scanner) advance() rune {
	r := s.readRune()
	if r != utf8.RuneError {
		s.current += 1
	}
	return r
}

func (s *Scanner) skipUntil(p func(rune) bool) {
	for {
		r := s.readRune()
		if r == utf8.RuneError || p(r) {
			break
		}
		s.current += 1
	}
}

func (s *Scanner) readRune() rune {
	return s.readRuneAhead(0)
}

func (s *Scanner) readRuneAhead(offset int) rune {
	offset += s.current
	for d := offset - len(s.chars) + 1; d > 0; d-- {
		r, sz, err := s.reader.ReadRune()
		if r == utf8.RuneError && sz == 1 {
			panic(errors.New("invalid UTF-8 sequence"))
		}
		if err != nil {
			if err == io.EOF {
				s.chars = append(s.chars, utf8.RuneError)
			} else {
				panic(err)
			}
		}
		s.chars = append(s.chars, r)
	}
	return s.chars[offset]
}

func (s *Scanner) mkToken(t TokenType) Token {
	return s.mkLiteral(t, nil)
}

func (s *Scanner) mkLiteral(t TokenType, literal interface{}) Token {
	lexeme := string(s.chars[:s.current])
	return Token{Type: t, Lexeme: lexeme, Literal: literal, Line: s.line}
}

var keywords = map[string]TokenType{
	"and":    AND,
	"class":  CLASS,
	"else":   ELSE,
	"false":  FALSE,
	"for":    FOR,
	"fun":    FUN,
	"if":     IF,
	"nil":    NIL,
	"or":     OR,
	"print":  PRINT,
	"return": RETURN,
	"super":  SUPER,
	"this":   THIS,
	"true":   TRUE,
	"var":    VAR,
	"while":  WHILE,
}
